File: src/qt/settingwindow.cpp
#include "settingwindow.h"
#include "ui_settingwindow.h"
#include<QDebug>
#include<stdlib.h>
#include <QDesktopWidget>
#include <QRect>
#include<QtGui>
SettingWindow::SettingWindow(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::SettingWindow)
{
    ui->setupUi(this);
    this->fresh();
    this->rawToRemove=-1;

}

SettingWindow::~SettingWindow()
{
    delete ui;
}






void SettingWindow::on_addButton_clicked()
{
    AddWindow *adw=new AddWindow;
    adw->set_sum(itemList.length());
    adw->show();
    this->connect(adw,SIGNAL(w_close()),this,SLOT(fresh()));

}

void SettingWindow::fresh()
{
    system("touch conf");
    ifstream confFile;
    confFile.open("conf",ios::in);
    listModel=new QStandardItemModel;
    itemList.clear();
    while(!confFile.eof())
    {
        string tmpstr;

        struct_items *item=new struct_items;
        confFile>>tmpstr;
        if("###"==tmpstr)
        {
            confFile>>tmpstr;
            confFile>>item->name;
            confFile>>tmpstr;
            confFile>>item->command;
            confFile>>tmpstr;
            confFile>>item->path;
            confFile>>tmpstr;
            confFile>>item->key;
        }
        if(item->path!="")
        {
            itemList.append(*item);
            QStandardItem *itemview=new QStandardItem(QIcon(\
            QString::fromLocal8Bit(item->path.c_str())),QString::fromLocal8Bit(item->name.c_str()));
            itemview->setEditable(false);
            listModel->appendRow(itemview);
            ui->listView->setModel(listModel);
        }
    }
}

void SettingWindow::on_okButton_clicked()
{

    this->close();
}

void SettingWindow::edit(QModelIndex index)
{

    AddWindow *aw=new AddWindow;
    aw->set_sum(itemList.length());
    aw->editmod();
    aw->show();
    aw->set_name(itemList.at(index.row()).name);
    aw->set_command(itemList.at(index.row()).command);
    aw->set_path(itemList.at(index.row()).path);
    aw->set_key(itemList.at(index.row()).key);
    this->rawToRemove=index.row();
    this->connect(aw,SIGNAL(w_close()),this,SLOT(fresh()));
    this->connect(aw,SIGNAL(edit_finish()),this,SLOT(remove_raw()));
}

void SettingWindow::remove_raw()
{
    itemList.removeAt(rawToRemove);
    listModel->removeRow(rawToRemove);
    ofstream outfile;
    outfile.open("conf",ios::out);
    int i;
    for(i=0;i<itemList.length();i++)
    {
        outfile<<"###"<<endl;
        outfile<<"[name] "<<itemList.at(i).name<<endl;
        outfile<<"[command] "<<itemList.at(i).command<<endl;
        outfile<<"[icon_path] "<<itemList.at(i).path<<endl;
        outfile<<"[key] "<<itemList.at(i).key<<endl;
    }

    outfile.close();
}

void SettingWindow::on_removeButton_clicked()
{
    int index=ui->listView->currentIndex().row();
    itemList.removeAt(index);
    listModel->removeRow(index);
    ofstream outfile;
    outfile.open("conf",ios::out);
    int i;
    for(i=0;i<itemList.length();i++)
    {
        outfile<<"###"<<endl;
        outfile<<"[name] "<<itemList.at(i).name<<endl;
        outfile<<"[command] "<<itemList.at(i).command<<endl;
        outfile<<"[icon_path] "<<itemList.at(i).path<<endl;
        outfile<<"[key] "<<itemList.at(i).key<<endl;
    }

    outfile.close();

}

void SettingWindow::on_listView_doubleClicked(const QModelIndex &index)
{
    this->edit(index);
}

File: bin/settingWindow.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

# 
# Author:     Ganger <ganger695988290@gmail.com>
# Maintainer: Ganger <ganger695988290@gmail.com>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from PyQt5.QtWidgets import QWidget,QListView,QPushButton
from PyQt5.QtGui import QStandardItemModel,QStandardItem,QIcon
from PyQt5.QtCore import  pyqtSignal, pyqtSlot
from addWindow import AddWindow
import cPickle

class SettingWindow(QWidget):
#	on_addButtonClicked=pyqtSignal()
#	on_removeButtonClicked=pyqtSignal()
#	on_okButtonClicked=pyqtSignal()
	finished=pyqtSignal()
	def __init__(self):
		QWidget.__init__(self)
		self.listview=QListView(self)
		self.addButton=QPushButton(self)
		self.removeButton=QPushButton(self)
		self.resize(630,440)
		self.okButton=QPushButton(self)
		self.listview.setGeometry(30,30,410,351)
		self.addButton.setGeometry(490,40,80,22)
		self.addButton.setText("add")
		self.addButton.clicked.connect(self.click_add)
		self.removeButton.setGeometry(490,80,80,22)
		self.removeButton.setText("remove")
		self.removeButton.clicked.connect(self.click_remove)
		self.okButton.setGeometry(490,150,80,22)
		self.okButton.setText("ok")
		self.okButton.clicked.connect(self.click_ok)
#		self.aw=null


		self.fresh()
	def click_ok(self):
		self.finished.emit()
		self.close()
	def click_add(self):
		self.aw=AddWindow()
		self.aw.show()
		self.aw.okSig.connect(self.fresh)
	def click_remove(self):
		self.remove()
	def fresh(self):
		confFile=open("conf","r")
		self.listModel=QStandardItemModel()
		self.itemList=cPickle.load(confFile)
		confFile.close()
		for  item in self.itemList:
			itemView=QStandardItem(QIcon(item.path),item.name)
			itemView.setEditable(False)
			self.listModel.appendRow(itemView)
			self.listview.setModel(self.listModel)
	def remove(self):
		index=self.listview.currentIndex().row()
		self.itemList.pop(index)
		self.listModel.removeRow(index)
		confFile=open("conf","w")
		cPickle.dump(self.itemList,confFile)
		confFile.close()
